crates/p2p/src/client/peer_agnostic/parse.rs

```diff
@@ -646,3 +654,69 @@ pub(crate) mod events {
         impl_take_parsed_and_should_stop!(events);
     }
 }
+
+/// Parses (header, signature) pairs. Expects chunks(2) as input,
+/// will panic if given a chunk with length not one or two.
+///
+/// Errors if the chunk is not a (header, signature) pair or a
+/// successful [BlockHeadersResponsePart::Fin].
+pub(crate) fn handle_signed_header_chunk(
+    chunk: Vec<BlockHeadersResponsePart>,
+) -> Option<anyhow::Result<SignedBlockHeader>> {
+    use anyhow::anyhow;
+
+    if let [single] = chunk.as_slice() {
+        match single {
+            Header(_) => {
+                return Some(Err(anyhow!("Stream finalized with header")));
+            }
+            Signatures(_) => {
+                return Some(Err(anyhow!("Stream finalized with signature")));
+            }
+            Fin(p2p_proto::common::Fin { error: Some(error) }) => {
+                return Some(Err(anyhow!("Stream finalized with error: {error:?}")));
+            }
+            Fin(_) => return None,
+        }
+    }
+
+    let [a, b] = chunk.as_slice() else {
+        panic!("Expected exactly two items in the chunk");
+    };
+
+    use BlockHeadersResponsePart::*;
+    let result = match (a, b) {
+        (Fin(_), _) | (_, Fin(_)) => Err(anyhow!("Received unexpected Fin")),
+        (Signatures(_), _) => Err(anyhow!("Received signature without header")),
+        (_, Header(_)) => Err(anyhow!("Received header without signature")),
```

Comment:
So this must cover (H,H) :thinking:

Reply:
Correct. Although this could move one line up and still be valid. We don't have to 100% identify the pairing. Its enough to say that we received a header that has no signature attached. Or vice versa.